cirq/circuits/text_diagram_drawer.py

```diff
@@ -59,12 +62,35 @@ class TextDiagramDrawer:
     """A utility class for creating simple text diagrams.
     """
 
-    def __init__(self):
-        self.entries = dict()  # type: Dict[Tuple[int, int], _DiagramText]
-        self.vertical_lines = []  # type: List[_VerticalLine]
-        self.horizontal_lines = []  # type: List[_HorizontalLine]
-        self.horizontal_padding = {}  # type: Dict[int, Union[int, float]]
-        self.vertical_padding = {}  # type: Dict[int, Union[int, float]]
+    def __init__(self,
+                 entries: Optional[Mapping[Tuple[int, int],
+                                           _DiagramText]] = None,
+                 horizontal_lines: Optional[Iterable[_HorizontalLine]] = None,
+                 vertical_lines: Optional[Iterable[_VerticalLine]] = None,
+                 horizontal_padding : Optional[Mapping[int, int]] = None,
+                 vertical_padding: Optional[Mapping[int, int]] = None,
+                 ) -> None:
+        self.entries = (dict() if entries is None else dict(entries)
+                ) # type: Dict[Tuple[int, int], _DiagramText]
+        self.horizontal_lines = ([] if horizontal_lines is None
+                else list(horizontal_lines)) # type: List[_HorizontalLine]
+        self.vertical_lines = ([] if vertical_lines is None
+                else list(vertical_lines)) # type: List[_VerticalLine]
+        self.horizontal_padding = (dict() if horizontal_padding is None
+                else dict(horizontal_padding)
+                ) # type: Dict[int, Union[int, float]]
+        self.vertical_padding = (dict() if vertical_padding is None
+                else dict(vertical_padding)
+                ) # type: Dict[int, Union[int, float]]
+
+    def __eq__(self, other):
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        attrs = ('entries', 'horizontal_lines', 'vertical_lines',
```

Comment:
Looks like this is a prime candidate for `@cirq.value_equality` (well, `@value.value_equality`).

Reply:
I completely agree, but when I try to use that I get circular import errors that I don't know how to fix.

Reply:
Might be a cycle via `cirq.protocols`... What's the stack trace?

Reply:
../../.virtualenvs/quantumlib/lib/python3.6/site-packages/py/_path/local.py:668: in pyimport
    __import__(modname)
cirq/__init__.py:21: in <module>
    from cirq.circuits import (
cirq/circuits/__init__.py:17: in <module>
    from cirq.circuits.text_diagram_drawer import (
cirq/circuits/text_diagram_drawer.py:21: in <module>
    from cirq import value
cirq/value/__init__.py:26: in <module>
    from cirq.value.periodic_value import (
cirq/value/periodic_value.py:19: in <module>
    import cirq.protocols
cirq/protocols/__init__.py:16: in <module>
    from cirq.protocols.apply_unitary import (
cirq/protocols/apply_unitary.py:23: in <module>
    from cirq import linalg
cirq/linalg/__init__.py:28: in <module>
    from cirq.linalg.decompositions import (
cirq/linalg/decompositions.py:267: in <module>
    @value.value_equality
E   AttributeError: module 'cirq.value' has no attribute 'value_equality'